File: 904.水果成篮.cpp
/*
 * @lc app=leetcode.cn id=904 lang=cpp
 *
 * [904] 水果成篮
 */

// @lc code=start
class Solution {
public:
  int totalFruit(vector<int> &fruits) {
    unordered_map<int, int> map;
    int left = 0, ans = 0;
    for (int right = 0; right < fruits.size(); ++right) {
      if (map.size() == 2 && map.find(fruits[right]) == map.end()) {
        auto key = min_element(
            map.begin(), map.end(),
            [](const auto &a, const auto &b) { return a.second < b.second; });
        left = key->second + 1;
        map.erase(key->first);
        map[fruits[right]] = right;
      } else {
        map[fruits[right]] = right;
      }
      ans = max(ans, right - left + 1);
    }
    return ans;
  }
};
// @lc code=end

File: 904.水果成篮.py
#
# @lc app=leetcode.cn id=904 lang=python3
#
# [904] 水果成篮
#

# @lc code=start
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        '''
        两个篮子则代表至多出现两个不同的数值
        动态滑动窗口，right端增大，当出现两个不同的数时，在第三个数停下
        出现第三个数，left端移动到第一个数的最后一个index+1位置处，right端继续增大
        计算窗口的最大值
        '''
        map={}
        ans=0
        left=right=0
        for right in range(len(fruits)): 
            if len(map)<=2 and fruits[right] not in map:
                left=map[fruits[left]]
                del map[fruits[left]]
                left=left+1
            else :
                map[fruits[right]]=right
            ans=max(ans,right-left+1)
        return ans            

# @lc code=end

